src/js/models/figure_model.js

```diff
@@ -742,53 +742,171 @@
+                while(c.x < right_x){
+                    next_panel = this.get_panel_at(c.x , c.y, sel);
+                    
+                    // if a panel doesn't exist at the current position c
+                    // just go to the next position until it reaches the selection boundaries
+                    if (next_panel) {
+                        row.push(next_panel);
+                        c = next_panel.get_centre();
+                    }else{
+                        next_panel = next_panel == undefined ? left : next_panel
+                    }
+                    c = {'x': c.x + next_panel.get('width'), 'y': c.y}
+                }
 
-                // if next_panel is not found, reached end of row. Try start new row...
-                if (typeof next_panel == 'undefined') {
+                // check that there is effecitvely a panel in the current row
+                if(row.length == 0){
+                    c = {'x': left_x + left.get('width')/2, 'y': c.y + left.get('height')}
+                }else{
+                    c = {'x': left_x + left.get('width')/2, 'y': c.y + row[0].get('height')}
                     grid.push(row);
-                    // next_panel is below the first of the current row
-                    c = row[0].get_centre();
-                    next_panel = this.get_panel_at(c.x, c.y + row[0].get('height'), sel);
-                    row = [];
-                }
-                if (next_panel) {
-                    row.push(next_panel);
-                }
+                }               
             }
 
-            var spacer = top_left.get('width')/20;
+            // get the row id of the most left panel
+            var left_panel_row = 0;
+            grid.forEach((row, i) => {
+                if (row.some(panel => panel.cid === left.cid)) {
+                    left_panel_row = i;
+                }
+            });
+            
+            // define the spacer between images
+            var spacer = left.get('width')/20;
             if (!isNaN(parseFloat(gridGap))) {
                 spacer = parseFloat(gridGap);
             }
-            var new_x = top_x,
-                new_y = top_y,
-                max_h = 0;
-            for (var r=0; r<grid.length; r++) {
-                row = grid[r];
+
+            var new_x = left_x,
+                new_y = left.get('y'),
+                max_h = 0,
+                ref_row = grid[left_panel_row],
+                last_panel_width = 0,
+                reference_grid = {};
+
+            // start aligning the row with the most left panel
+            // This row is considered as a reference to align
+            // the rest of the panels
+            var global_index = 0
+            for (var c = 0; c < ref_row.length; c++) {
+                let panel = ref_row[c];
+                var current_x = panel.get('x')
+                var gap = Math.floor(Math.abs(current_x - new_x) / last_panel_width)
```

Comment:
On the first iteration through this for loop, `last_panel_width` is `0`.
If `current_x` and `new_x` are equal then `gap` is `NaN` and the inner for-loop below doesn't get executed, but if `current_x` and `new_x` are not equal then `gap` is `Infinite` and the inner for loop below goes on "forever".

I came across this bug when I randomly tried to put 2 overlapping panels into a grid:

<img width="326" height="373" alt="Screenshot 2025-12-18 at 15 44 15" src="https://github.com/user-attachments/assets/37c10536-a182-4f63-a71a-009f62fc68d5" />

Reply:
@Rdornier - you only see this bug if the left-most panel is lower than the other one.
I think we just need to skip a bunch of logic for the first iteration through the loop?